Misc READEME.md cleanup (#31)
WICG/scheduling-apis: README.md

times of contention, which increases task latency and can result in degraded
quality of experience.

Consider a "search-as-you-type" application. This app needs to be responsive to
user input, i.e. users typing in the search-box. At the same time, any
animations on the page must be rendered smoothly, and the work for fetching and
preparing search results and updating the page must also progress quickly.
There are a lot of different deadlines to meet for the app developer. It is
easy for any long running script work to hold up the main thread and cause
responsiveness issues for typing, rendering animations, or updating search
results.

Another example pinch-zooming in a map application. The app needs to
continuously respond to the input, update the rendering, and potentially fetch
new content to be displayed. Similar to the search-as-you-type example, long
running script work could block other tasks, making the application feel laggy.

## Current Solutions, Their Limitations, and APIs to Fill the Gaps

Dealing with contention is largely a scheduling problem: to the degree that
work can be reordered in an more optimal way, scheduling can have a positive

> impact. What makes this problem more pronounced on the web is that tasks run to
> completion, meaning the UA cannot interrupt a task to run high priority work
> like processing user input. This problem is generally tackled in userspace by

Reviewer: ```suggestion
impact. This problem is more pronounced on the web because tasks run to
completion: the UA cannot preempt a task to run higher priority work
like processing user input. This problem is generally tackled in userspace by
```

Dunno if "preempt" is actually better than "interrupt"; it's just the term I expect to see in this space.

these chunks of work&mdash;prioritizing and executing work async at an
appropriate time relative to current situation of user and browser. And while
userspace schedulers have been effective in improving responsiveness, there are
several problems they still face:

 1. **Determining when to yield to the browser**: yielding has overhead&mdash;the
    overhead of posting a task and context switching, the cost of regaining
    control, etc. This can lead to increased task latency.

    Making intelligent decisions about when to yield is difficult with limited
    knowledge. Scheduling primitives can help userspace schedulers make better
    decisions, e.g. [isInputPending()](https://github.com/WICG/is-input-pending)
    and [isFramePending()](https://github.com/szager-chromium/isFramePending/blob/master/explainer.md).

 2. **Regaining control after yielding**: chunking work and yielding is
    necessary for improving responsiveness, but it comes at a cost: when
    yielding to the event loop, a task that yields has no way to *continue*
    without arbitrary work of the *same priority* running first, e.g. other

>     script. This disincentivizes yielding for script that requires low task

Reviewer: ```suggestion
    script. This disincentivizes yielding from a script that requires low task
```

    latency. Providing a primitive like [scheduler.yield()](./explainers/yield-and-continuation.md)
    that is designed to take into account this async userspace task model can
    help, as the scheduler can prioritize these continuations more fairly.

>  3. **Coordination between (cooperating) actors**: Most userspace schedulers

Reviewer: Consider making this the first item so people don't ask why we're doing the third-most-important thing first.

>     This disparate set of scheduling APIs makes it onerous for developers to

Reviewer: https://www.thesaurus.com/browse/onerous ...

```suggestion
    This disparate set of scheduling APIs makes it difficult for developers to
```

?

>     write scheduling code, and requires expert knowledge of the Browser's event

Reviewer: Strictly,

```suggestion
    write scheduling code and requires expert knowledge of the browser's event
```

>     loop to do so. Creating a unified native scheduling API can alleviate this,
>     which we are also addressing with the [scheduler.postTask() API](./explainers/prioritized-post-task.md).

Reviewer: ```suggestion
    loop to do so. Creating a unified native scheduling API—[scheduler.postTask()](./explainers/prioritized-post-task.md)—will alleviate this.
```

Author: Nice. Done.

Some of the other APIs we are considering in this space are noted [here](./misc/low-level-apis.md).

## APIs and Status

 | API | Abstract | Status | Links |
 | --- | --- | --- | --- |
 | `scheduler.postTask()` | An API for scheduling and controlling prioritizing tasks. | An origin trial recently concluded in Chrome M89, and we are currently working towards shipping the API. The feature is available [behind a flag](./origin-trial-status.md) in Chromium. | [Explainer](./explainers/prioritized-post-task.md) [Spec](https://wicg.github.io/scheduling-apis/) |
 | `scheduler.yield()` | An API for breaking up long tasks by yielding to the browser, continuing after being rescheduled by the scheduler. | This feature is actively being designed, and the explainer will be updated soon with our current thinking. | [Explainer](./explainers/yield-and-continuation.md) |
 | `scheduler.wait()` | This enables tasks to yield and resume after some amount of time, or perhaps after an event has occurred. | This feature is currently being co-designed with `scheduler.yield()`. | [Related Discussion](https://github.com/WICG/scheduling-apis/issues/19) |

>  | `scheduler.currentTaskSignal` | This API provides a way to get the currently running task's `TaskSignal`, which can be used to schedule dependent tasks. | This API is available behind a flag in Chromium, but we are currently re-evaluating the API based on the greater notion of a *userspace task* in context of `scheduler.yield()`. | [Explainer](./explainers/post-task-propagation.md) |

Reviewer: "greater" is a little confusing. "Larger"? Or just "notion"? You don't have the "userspace task" essay written yet, right?

>  | `scheduler.currentTaskSignal` | This API provides a way to get the currently running task's `TaskSignal`, which can be used to schedule dependent tasks. | This API is available behind a flag in Chromium, but we are currently re-evaluating the API based on the greater notion of a *userspace task* in context of `scheduler.yield()`. | [Explainer](./explainers/post-task-propagation.md) |

Reviewer: ```suggestion
 | `scheduler.currentTaskSignal` | This API provides a way to get the currently running task's `TaskSignal`, which can be used to schedule dependent tasks. | This API is available behind a flag in Chromium, but we are currently re-evaluating the API based on the greater notion of a *userspace task* in the context of `scheduler.yield()`. | [Explainer](./explainers/post-task-propagation.md) |
```
